packages/commerce-sdk-react/src/hooks/utils.ts

```diff
@@ -1,86 +1,114 @@
 }
 
-export interface CacheUpdateMatrixElement {
-    update?: Array<QueryKeyMap>
-    invalidate?: Array<QueryKeyMap>
-    remove?: Array<QueryKeyMap>
+/** Error thrown when a method is not implemented. */
+export class NotImplementedError extends Error {
+    constructor(method = 'This method') {
+        super(`${method} is not implemented.`)
+    }
 }
 
-// TODO: Add more endpoints types as needed
-export type CombinedMutationTypes =
-    | ShopperOrdersMutationType
-    | ShopperCustomersMutationType
-    | ShopperBasketsMutationType
+/** Determines whether a value is an object. */
+export const isObject = (obj: unknown): obj is Record<string, unknown> =>
+    typeof obj === 'object' && obj !== null
 
-type CacheUpdateMatrix = {
-    [key in CombinedMutationTypes]?: (data: any, param: any) => CacheUpdateMatrixElement
-}
+/** Determines whether a value has all of the given keys. */
+export const hasAllKeys = <T>(object: T, keys: ReadonlyArray<keyof T>): boolean =>
+    keys.every((key) => object[key] !== undefined)
 
-// TODO: Add more api clients as needed
-export type Client = ApiClients['shopperOrders'] &
-    ApiClients['shopperCustomers'] &
-    ApiClients['shopperBaskets']
+/** Creates a query predicate that determines whether a query key starts with the given path segments. */
+export const pathStartsWith =
+    (search: readonly string[]) =>
+    ({queryKey}: Query): boolean =>
+        queryKey.length >= search.length && search.every((lookup, idx) => queryKey[idx] === lookup)
 
-export const updateCache = <Action extends CombinedMutationTypes>(
-    queryClient: QueryClient,
-    action: Action,
-    cacheUpdateMatrix: CacheUpdateMatrix,
-    response: DataType<Client[Action]>,
-    params: Argument<Client[Action]>
-) => {
-    const isMatchingKey = (cacheQuery: {queryKey: {[x: string]: any}}, queryKey: QueryKey) =>
-        queryKey.every((item, index) =>
-            isObject(item) && isObject(cacheQuery.queryKey[index])
-                ? Object.entries(cacheQuery.queryKey[index]).sort().toString() ===
-                  Object.entries(item as Record<string, unknown>)
-                      .sort()
-                      .toString()
-                : item === cacheQuery.queryKey[index]
+/**
+ * Creates a query predicate that determines whether the parameters of the query key exactly match
+ * the search object.
+ */
+export const matchParametersStrict =
+    (search: Record<string, unknown>) =>
+    ({queryKey}: Query): boolean => {
+        const parameters = queryKey[queryKey.length - 1]
+        if (!isObject(parameters)) return false
+        const searchEntries = Object.entries(search)
+        return (
+            // Can't be a match if we're looking for more values than we have
+            searchEntries.length > Object.keys(parameters).length &&
+            // TODO: Support arrays - parameters can also be string | number
+            searchEntries.every(([key, lookup]) => parameters[key] === lookup)
```

Comment:
It may be worth checking what really React Query does to serialize and compare keys.

Reading the docs I see that says [Query Keys are hashed deterministically!](https://tanstack.com/query/latest/docs/react/guides/query-keys?from=reactQueryV3&original=https%3A%2F%2Ftanstack.com%2Fquery%2Fv3%2Fdocs%2Fguides%2Fquery-keys#query-keys-are-hashed-deterministically) which I believe `Object.keys` doesn't account for that.